chacha20_poly1305/src/poly1305.rs

```diff
@@ -0,0 +1,242 @@
+        let r1 = (u32::from_le_bytes([key[3], key[4], key[5], key[6]]) >> 2) & 0x03ffff03;
+        let r2 = (u32::from_le_bytes([key[6], key[7], key[8], key[9]]) >> 4) & 0x03ffc0ff;
+        let r3 = (u32::from_le_bytes([key[9], key[10], key[11], key[12]]) >> 6) & 0x03f03fff;
+        let r4 = (u32::from_le_bytes([key[12], key[13], key[14], key[15]]) >> 8) & 0x000fffff;
+
+        let s0 = u32::from_le_bytes([key[16], key[17], key[18], key[19]]);
+        let s1 = u32::from_le_bytes([key[20], key[21], key[22], key[23]]);
+        let s2 = u32::from_le_bytes([key[24], key[25], key[26], key[27]]);
+        let s3 = u32::from_le_bytes([key[28], key[29], key[30], key[31]]);
+
+        Poly1305 {
+            r: [r0, r1, r2, r3, r4],
+            s: [s0, s1, s2, s3],
+            acc: [0; 5],
+            leftovers: [0; 16],
+            leftovers_len: 0,
+        }
+    }
+
+    /// Add message to be authenticated, can be called multiple times before creating tag.
+    pub fn input(&mut self, message: &[u8]) {
+        // Process previous leftovers if the message is long enough to fill the leftovers buffer. If
+        // the message is too short then it will just be added to the leftovers at the end. Now if there
+        // are no leftovers, but the message can fill the buffer, it will process that buffer and
+        // and process the rest of the message later on.
+        let fill =
+            if self.leftovers_len + message.len() >= 16 { 16 - self.leftovers_len } else { 0 };
+
+        if fill > 0 {
+            self.leftovers[self.leftovers_len..].copy_from_slice(&message[0..fill]);
+
+            let msg_slice = prepare_padded_message_slice(&self.leftovers, false);
+            for (i, b) in msg_slice.iter().enumerate() {
+                self.acc[i] += *b;
+            }
+            self.r_times_a();
+            self.leftovers_len = 0;
+        }
+
+        // Remove prefix already processed in leftovers.
+        let remaining_message = &message[fill..];
+
+        // Add message to accumulator.
+        let mut i = 0;
+        while i < remaining_message.len() / 16 {
+            let msg_slice =
+                prepare_padded_message_slice(&remaining_message[i * 16..(i + 1) * 16], false);
+            for (i, b) in msg_slice.iter().enumerate() {
+                self.acc[i] += *b;
+            }
+            self.r_times_a();
+            i += 1;
+        }
+
+        // Save any leftovers.
+        if remaining_message.len() % 16 > 0 {
+            let message_index = remaining_message.len() - (remaining_message.len() % 16);
+            let new_len = self.leftovers_len + remaining_message.len() % 16;
+            self.leftovers[self.leftovers_len..new_len]
+                .copy_from_slice(&remaining_message[message_index..]);
+            self.leftovers_len = new_len;
+        }
+    }
+
+    /// Generate authentication tag.
+    pub fn tag(&mut self) -> [u8; 16] {
```

Comment:
Since this is basically finalizing the hash, it should probably take `self`, right?

Reply:
Yea, that makes sense.